gym/vector/async_vector_env.py

```diff
@@ -0,0 +1,482 @@
+            self.observations = create_empty_array(
+                self.single_observation_space, n=self.num_envs, fn=np.empty)
+
+        self.parent_pipes, self.processes = [], []
+        self.error_queue = self.ctx.Queue()
+        with clear_mpi_env_vars():
+            for index in range(self.num_envs):
+                process, parent_pipe = self._start_process(index)
+                self.parent_pipes.append(parent_pipe)
+                self.processes.append(process)
+
+        self.waiting_reset = False
+        self.waiting_step = False
+        self._check_observation_spaces()
+
+    def seed(self, seeds=None):
+        self._assert_is_running()
+        if seeds is None:
+            seeds = [None for _ in range(self.num_envs)]
+        if isinstance(seeds, int):
+            seeds = [seeds + i for i in range(self.num_envs)]
+        assert len(seeds) == self.num_envs
+
+        for pipe, seed in zip(self.parent_pipes, seeds):
+            pipe.send(('seed', seed))
+        for pipe in self.parent_pipes:
+            pipe.recv()
+
+    def reset_async(self):
+        self._assert_is_running()
+        if self.waiting_step:
+            logger.warn('Calling `reset_async` while waiting for a pending '
+                'call to `step` to complete. Cancelling the previous call to '
+                '`step`...')
+            try:
+                self.step_wait(timeout=0)
+            except mp.TimeoutError:
+                pass
+
+        if self.waiting_reset:
+            logger.error('Calling `reset_async` while waiting for a pending '
+                'call to `reset` to complete. Closing `{0}`...'.format(self))
+            self.close(terminate=True)
+            raise AlreadyResettingError('Calling `reset_async` while waiting '
+                'for a pending call to `reset` to complete.')
+
+        for pipe in self.parent_pipes:
+            pipe.send(('reset', None))
+        self.waiting_reset = True
+
+    def reset_wait(self, timeout=None):
+        """
+        Parameters
+        ----------
+        timeout : int, optional
+            Number of seconds before the call to `reset_wait` times out. If
+            `None`, the call to `reset_wait` never times out.
+
+        Returns
+        -------
+        observations : sample from `observation_space`
+            A batch of observations from the vectorized environment.
+        """
+        self._assert_is_running()
+        if self.waiting_step or (not self.waiting_reset):
+            logger.error('Calling `reset_wait` without any prior call to '
+                '`reset_async`. Closing `{0}`...'.format(type(self).__name__))
+            self.close(terminate=True)
+            raise NotResettingError('Calling `reset_wait` without any prior '
+                'call to `reset_async`.')
+
+        if not self.poll(timeout):
+            self.waiting_reset = False
+            raise mp.TimeoutError('The call to `reset_wait` has timed out after '
```

Comment:
If this is going to use TimeoutError, why not use the python builtin instead of the multiprocessing one?

Reply:
For Python 2 compatibility, Python 2 does not have `TimeoutError`